---
title: "SimulationsNatGen"
author: "Luis Zapata"
date: "9/3/2022"
output:
  html_document: default
  pdf_document: default
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(ggpubr)
library(tidyverse)
library(crayon)
library(pio)
library(easypar)
library(parallel)
rm(list=ls())
```

### Load simulations

```{r load_simulations_100}
load("~/tools/dNdSSimulator/example/Zapata22_NatGenExample.Rdata")
```

### Default parameters

```{r detail_simulations, eval=F}
######### Number of repetitions #########
TASKS = 1

######## Define our set of parameters for the SIESTA model ###########
params = NULL

####Define initial population size
params['n0'] <- 5

######## Global parameters #########
params['GENERATIONS'] = 100
params['Kcapacity'] = 2000

#######Set probability of cell survival
params['ps'] <- 0.5
params['pdiff'] <- (1 - params['ps'])

######Set mutation occurrence parameters
params['u'] = 1 /  (10 ^ 6) #muts/ p divisionm / p bp, ##mutation occurrence/ polymerase error
params['L'] = 50 * (10 ^ 6) ###Length of the genome/exome size of exome syn plus nonsyn mutations
params['C'] = 1 /  (10 ^ 2) #correction efficiency, it allows only 1 out of X number of mutations to be preserved at each cell division

#######Set mutation probabilities
params['pnad'] <- 0.01 #Nonsyn - driver mutations (increase proliferation rate, oncogenes/ reduce correction efficiency, mut_modifiers, decrease death rate)
params['pnai'] <- 0.10 #Nonsyn - deleterious mutations (immunogenic mutations)
params['pnak'] <- 0.00 #Nonsyn - deleterious mutations (death)
params['pnae'] <- 0.01 #Nonsyn - escape immune system mutations
params['pnsi'] <- params['pnai']/3    # Synonymous mutations in imunopeptidome
params['pnsd'] <- params['pnad']/3    # Synonymous mutations in driver
params['pnap'] <- 0.75 - ( params['pnad'] +  params['pnai'] +  params['pnae'] +  params['pnak']) #Nonsyn - passenger mutations
params['pns']  <- 0.25 - ( params['pnsi'] +  params['pnsd']) #Synonymous mutations

```

### Probability of immune activity set to ACTIVE (pattack=1) or INACTIVE (pattack=0)

```{r immune_response, eval= F}
####Set value for the probability of the immune system to kill an immunogenic population with probability 0 = INACTIVE
params['pattack'] = 0

####Set value for the probability of the immune system to kill an immunogenic population with probability 0 = ACTIVE
params['pattack'] = 1
```


### Set minimum allele frequency for mutations

```{r analys_freq2}
dnds_IA_100<-calc_freq_dnds(RUNS_IA_100,RUNS_IA_100_genotypes, freq = 0.01) %>% set_params(.,mod = "A",imm = "ACTIVE",esc = 0)

dnds_IA_0<-calc_freq_dnds(RUNS_IA_0,RUNS_IA_0_genotypes, freq = 0.01) %>% set_params(.,mod = "A",imm = "INACTIVE",esc = 0)

dnds_both<-bind_rows(dnds_IA_100,dnds_IA_0)
```

### Plot dN/dS distribution for simulations > 1000 cells

```{r plot_individuals,message=F,cache.comments=F}

supp_fig1<-ggstatsplot::ggbetweenstats(dnds_both %>% filter(pop>1000), x = "immune", y = "dnds_immune",plot.type = "box", type = "np",ylab = "Immune dN/dS", xlab = "Immune system")
supp_fig1

```
